PATCH-P1-FIX-UPSTREAM: Fix heap overflow vulnerability in regex library
BUGS: bnc#922043

Description of the vulnerability from
https://guidovranken.wordpress.com/2015/02/04/full-disclosure-heap-overflow-in-h-spencers-regex-library-on-32-bit-systems/
"Fix heap overflow vulnerability in Henry Spencer’s regex library,
affecting 32 bit systems only. Variable ‘len’ is here enlarged to such 
an extent that, in the process of enlarging (multiplication and addition),
causes the 32 bit register/variable to overflow." 

Maintainer: Kristyna Streitova <kstreitova@suse.com>

Index: mysql-5.5.43/regex/regcomp.c
===================================================================
--- mysql-5.5.43.orig/regex/regcomp.c
+++ mysql-5.5.43/regex/regcomp.c
@@ -138,7 +138,15 @@ CHARSET_INFO *charset;
 							(NC-1)*sizeof(cat_t));
 	if (g == NULL)
 		return(MY_REG_ESPACE);
-	p->ssize = (long) (len/(size_t)2*(size_t)3 + (size_t)1); /* ugh */
+	{
+	  /* Patched for CERT Vulnerability Note VU#695940, Feb 2015. */
+	  size_t new_ssize = len/(size_t)2*(size_t)3 + (size_t)1; /* ugh */
+	  if (new_ssize < len || new_ssize > LONG_MAX / sizeof(sop)) {
+	    free((char *) g);
+	    return MY_REG_INVARG;
+	  }
+	  p->ssize = (long) new_ssize;
+	}	
 	p->strip = (sop *)malloc(p->ssize * sizeof(sop));
 	p->slen = 0;
 	if (p->strip == NULL) {
